packages/artifact/src/internal/download-http-client.ts

```diff
@@ -0,0 +1,195 @@
+  /**
+   * Fetches a set of container items that describe the contents of an artifact
+   * @param artifactName the name of the artifact
+   * @param containerUrl the artifact container URL for the run
+   */
+  async getContainerItems(
+    artifactName: string,
+    containerUrl: string
+  ): Promise<QueryArtifactResponse> {
+    // the itemPath search parameter controls which containers will be returned
+    const resourceUrl = new URL(containerUrl)
+    resourceUrl.searchParams.append('itemPath', artifactName)
+
+    // no concurrent calls so a single httpClient without the http-manager is sufficient
+    const client = createHttpClient()
+
+    // no keep-alive header, client disposal is not necessary
+    const requestOptions = getRequestOptions('application/json')
+    const rawResponse = await client.get(resourceUrl.toString(), requestOptions)
+    const body: string = await rawResponse.readBody()
+    if (isSuccessStatusCode(rawResponse.message.statusCode) && body) {
+      return JSON.parse(body)
+    }
+    // eslint-disable-next-line no-console
+    console.log(rawResponse)
+    throw new Error(`Unable to get ContainersItems from ${resourceUrl}`)
+  }
+
+  /**
+   * Concurrently downloads all the files that are part of an artifact
+   * @param downloadItems information about what items to download and where to save them
+   */
+  async downloadSingleArtifact(downloadItems: DownloadItem[]): Promise<void> {
+    const DOWNLOAD_CONCURRENCY = getDownloadFileConcurrency()
+    // limit the number of files downloaded at a single time
+    const parallelDownloads = [...new Array(DOWNLOAD_CONCURRENCY).keys()]
+    let downloadedFiles = 0
+    await Promise.all(
+      parallelDownloads.map(async index => {
+        while (downloadedFiles < downloadItems.length) {
+          const currentFileToDownload = downloadItems[downloadedFiles]
+          downloadedFiles += 1
+          await this.downloadIndividualFile(
+            index,
+            currentFileToDownload.sourceLocation,
+            currentFileToDownload.targetPath
+          )
+        }
+      })
+    )
+
+    // done downloading, safety dispose all connections
+    this.downloadHttpManager.disposeAndReplaceAllClients()
+  }
+
+  /**
+   * Downloads an individual file
+   * @param httpClientIndex the index of the http client that is used to make all of the calls
+   * @param artifactLocation origin location where a file will be downloaded from
+   * @param downloadPath destination location for the file being downloaded
+   */
+  private async downloadIndividualFile(
+    httpClientIndex: number,
+    artifactLocation: string,
+    downloadPath: string
+  ): Promise<void> {
+    const stream = fs.createWriteStream(downloadPath)
+    const client = this.downloadHttpManager.getClient(httpClientIndex)
+    const requestOptions = getRequestOptions('application/octet-stream', true)
+    const response = await client.get(artifactLocation, requestOptions)
+    await response.readBody()
```

Comment:
Does this load the entire file into memory?  Do we need this?

Reply:
I added the `response.readBody` as a precaution to make sure the connection gets closed so that there is no resource leak. Similar to what was causing the weird 1 minute wait with upload after everything was one.

I tested with and without `response.readBody()` and it both cases it's working fine for download so I'll remove it. In the initial version of the package I don't have this snippet and it works fine so it looks like only upload needs to read the body.